store/postgres/src/store.rs

```diff
@@ -665,17 +704,142 @@ impl Store {
+    ///   spawn the database work in a thread though, this expectation no longer
+    ///   holds because the spawned task is the independent of this future. So,
+    ///   this method provides a cancel token which indicates that the `Future`
+    ///   has been dropped. This isn't *quite* as good as drop on cancel,
+    ///   because a drop on cancel can do things like cancel http requests that
+    ///   are in flight, but checking for cancel periodically is a significant
+    ///   improvement.
+    ///
+    ///   The implementation of the supplied closure should check for cancel
+    ///   between every operation that is potentially blocking. This includes
+    ///   any method which may interact with the database. The check can be
+    ///   conveniently written as `token.check_cancel()?;`. It is low overhead
+    ///   to check for cancel, so when in doubt it is better to have too many
+    ///   checks than too few.
+    ///
+    /// # Panics:
+    ///   * This task will panic if the supplied closure panics
+    ///   * This task will panic if the supplied closure returns Err(Cancelled)
+    ///     when the supplied cancel token is not cancelled.
+    async fn with_conn<T: Send + 'static>(
+        &self,
+        f: impl 'static
+            + Send
+            + FnOnce(
+                &PooledConnection<ConnectionManager<PgConnection>>,
+                &CancelHandle,
+            ) -> Result<T, CancelableError>,
+    ) -> Result<T, Error> {
+        let _permit = CONNECTION_LIMITER.acquire().await;
+        let store = self.clone();
+
+        let cancel_guard = CancelGuard::new();
+        let cancel_handle = cancel_guard.handle();
+
+        let result = spawn_blocking_async_allow_panic(move || {
+            // It is possible time has passed between scheduling on the
+            // threadpool and being executed. Time to check for cancel.
+            cancel_handle.check_cancel()?;
+
+            // A failure to establish a connection is propagated as though the
+            // closure failed.
+            let conn = store.get_conn()?;
+
+            // It is possible time has passed while establishing a connection.
+            // Time to check for cancel.
+            cancel_handle.check_cancel()?;
+
+            f(&conn, &cancel_handle)
+        })
+        .await;
+
+        drop(cancel_guard);
+
+        // Finding cancel isn't technically unreachable, since there is nothing
+        // stopping the supplied closure from returning Canceled even if the
+        // supplied handle wasn't canceled. That would be very unexpected, the
+        // doc comment for this function says we will panic in this scenario.
+        match result {
+            Ok(t) => Ok(t),
+            Err(CancelableError::Error(e)) => Err(e),
+            Err(CancelableError::Cancel) => panic!("The closure supplied to with_entity_conn must not return Err(Canceled) unless the supplied token was canceled."),
```

Comment:
The cancel token supplied to the task closure is only canceled if this future is dropped while we were waiting for the task to finish. If we reached this point, the supplied closure must have finished executing before the cancel token was canceled. If the value is then `CancelableError::Cancel` the method decided to return `CancelableError::Cancel` even though cancellation was not requested. In that case, there is no good return value for `with_conn` since it does not return `CancelableError` (nor should it, since it follows the better drop-to-cancel pattern).